### DISCUSSION TITLE
Changing border width and color

### DISCUSSION THREAD

**Discussion by KillTheMule** [fltk-rs/fltk-rs#1022] [Q&A] [answered]
The widgets `Output` and `Choice` come with a border around them. Is there a way to change the width and the color of these borders?

**Comment by MoAlyousef** [comment, marked as answer]
Regarding Output, the default frame type is a DownBox. You can change it using the set_frame method (available frame types are in 
 enum::FrameType), and setting the color. FrameTypes which have Frame in them are hollow, while those with Box in them are not.
`out.set_frame(FrameType::BorderFrame); out.set_color(Color::Red);`

Choice doesn't expose the underlying widget which has a down box, so it can only be changed by either changing the app scheme or by globally changing the draw function of DownBox.
```rust
use fltk::{enums::*, prelude::*, *};

fn my_down_box(x: i32, y: i32, w: i32, h: i32, col: Color) {
    draw::draw_rect_fill(x, y, w, h, Color::Red);
    draw::draw_rect_fill(x + 1, y + 1, w - 2, h - 2, Color::BackGround2); // change values to change thickness
}

fn main() {
    let a = app::App::default();
    app::set_frame_type_cb(FrameType::DownBox, my_down_box, 0, 0, 0, 0);
    let mut win = window::Window::new(100, 100, 400, 300, None);
    win.set_color(Color::from_rgb(211, 211, 211));
    let mut inp = input::Input::new(50, 10, 100, 30, None); // would work for any widget which has a DownBox frame type
    let mut choice = menu::Choice::new(50, 100, 100, 30, None);
    choice.add_choice("Choice 1| Choice 2| choice 3");
    win.end();
    win.show();
    a.run().unwrap();
}
```

The other option is to implement your own custom choice widget (the custom_popup example might offer some ideas)

**Reply by KillTheMule** [reply]
That all worked very well, thank you! I can't believe I didn't try `set_frame` on the `Output`.

Can I follow up with a question? The box for the `Choice` menu can be styled the way you showed me, but the menu itself (the one that unfolds after clicking on the box) is just plain white, which makes it kinda hard on the eyes, because there's a lot of white in my app :) Is there a way to make it a different color than the window color (it seems this is what it uses)? Or, even better, can I somehow get a border around it? Thanks in advance!

Speaking of thanks, you don't have set up something to donate, or did I miss it? I'm a fan of your work, and you're being so helpfull, I'd like to give something back. Would you be interested in PR's that make little improvements to the docs, in the "this would have helped me ask less questions" sense?

**Reply by MoAlyousef** [reply]
Thank you. I haven't set up donations or anything. But I do appreciate little improvements to the docs :)

The color of the poped up menu window itself is the same as the BackGround2 color. This can be changed using `app::set_background2_color(r, g, b)`.
```rust
use fltk::{prelude::*, enums::*, *};

fn main() {
    let a = app::App::default();
    app::set_background2_color(255, 199, 200);
    let mut win = window::Window::default().with_size(400, 300);
    let mut choice = menu::Choice::new(160, 200, 80, 30, "Click");
    choice.add_choice("Choice 1|Choice 2");
    choice.set_frame(FrameType::BorderBox);
    win.end();
    win.show();
    a.run().unwrap();
}
```
You can add a border using the set_frame method of the choice, it actually works on the popped up menu window.

I've also uploaded a [custom_choice](https://github.com/fltk-rs/fltk-rs/blob/master/fltk/examples/custom_choice.rs) example, it allows custom styling of each of the element widgets inside it.

**Reply by KillTheMule** [reply]
Ok, that was kinda weird. I did try `set_frame` on `Choice`, but it didn't seem to to anything, because I used `BorderBox` for it, which just looks like `FlatBox` here. Using `EngravedBox` produces the visual I'd want. Thanks a bunch!

I'll see if I can cook up some doc improvements once I shipped my app :+1:
